purchase_invoice.py

```diff
@@ -1178,73 +378,73 @@ def store_json(json_data: dict):
+    def parse_invoice(self, invoice_json, infile, account=None, paid_by_submitter=False, given_supplier=None,
+                      is_test=False, check_dup=True, manual_edit=False):
+        normal_purchase_data = None
+        google_purchase_data = None
+        if invoice_json:
+            print("Nutze Google invoice parser")
+            purchase_invoice_google_parser = PurchaseInvoiceGoogleParser(self, invoice_json, given_supplier, is_test)
+            purchase_invoice_google_parser.set_purchase_info()
+            google_purchase_data = purchase_invoice_google_parser.get_purchase_data()
+        print("Nutze internen Parser")
+        self.extract_items = False
+        lines = pdf_to_text(infile)
         try:
-            self.extract_main_info(invoice_json, given_supplier, infile)
-            self.compute_total()
+            if lines:
+                head = lines[0][0:140]
+                if not head[0:10].split():
+                    for line in lines[0:10]:
+                        if self.company.name != 'Bremer SolidarStrom' and len(line) > 2 and (
+                                line[-2] == '£' or line[-3] == '£'):
+                            head = "Kornkraft Naturkost GmbH"
+                            break
+                supps = dict(PurchaseInvoice.suppliers)
+                if self.company.name != 'Laden':
+                    del supps['Rechnung']
+                for supplier, info in supps.items():
+                    if supplier in head:
+                        self.parser = supplier
+                        self.extract_items = self.update_stock
+                        if info['raw']:
+                            self.raw = True
+                            lines = pdf_to_text(infile, True)
+                        if 'supplier' in info:
+                            self.supplier = info['supplier']
+                        else:
+                            self.supplier = supplier
+                        print("Verwende Rechnungsparser für ", self.supplier)
+                        if given_supplier and self.supplier != given_supplier:
+                            print("abweichend von PreRechnung: ", given_supplier)
+                        self.multi = info['multi']
+                        purchase_invoice_parser = PurchaseInvoiceParser(self, info['parser'], lines)
+                        purchase_invoice_parser.set_purchase_info()
+                        normal_purchase_data = purchase_invoice_parser.get_purchase_data()
         except Exception as e:
-            if self.update_stock:
+            if google_purchase_data:
+                print(e)
+            elif self.update_stock:
                 raise e
             elif not is_test:
                 print(e)
                 print("Rückfall auf Standard-Rechnungsbehandlung")
+        if google_purchase_data:
+            if normal_purchase_data:
+                google_purchase_data.update(normal_purchase_data)
```

Comment:
this overwrites existing data. In case of unstructured data like bill_no, this is OK. However, in case of structured data, like the list of items, some more complex merging should be done. (by the way:  apply_info_changes seems to do a more complex merge?)

Reply:
I understand what your concern is, but I can't think of a clean solution to merge the items of two jsons .